### DISCUSSION TITLE
space in command or subcommand

### DISCUSSION THREAD

**Discussion by achursin1986** [daniele77/cli#205] [Ideas / Feature requests]
Hi,

thanks for the great lib!

just an idea, it was mentioned "Cisco-like" in the project description, however is there a way to implement a command like : 
"show database" with space? I have tried a literal it's accepted on compile but doesn't work saying wrong command. Like in Cisco world there are commands with a common prefix :
show ip interface
show ip route 
Is there a way to implement the same? 
Sorry if it was asked before but I could not find it.

**Comment by daniele77** [comment]
Hi @achursin1986 , thank you for using the library.
You can get that behavior by using menus. If you set up a menu structure like this:

    show         # menu
      |---- ip     # menu
              |----- interface   # command
              |----- route        # command

with that menus structure, from the root menu, you can then input the following commands:

    show ip interface
    show ip route

because the library shell allows you to enter the "full path" of a command, w/o entering explicitly in the submenus.
You can check this feature by starting the example `complete.cpp` coming with the library entering the command:

    > sub subsub hello

**Reply by achursin1986** [reply]
Thanks for your prompt response @daniele77 , indeed just checked complete example. Might be a small followup: can we then block access to submenus , so "show ip route" works but we can't get into just "show" or "show ip" ?

**Comment by daniele77** [comment]
I don't see a great value in forbidding the menu access.
Can you provide a use case where this feature is essential? In this case, we can consider adding the feature to the roadmap.

**Comment by achursin1986** [comment]
yes, it can be handy to implement show commands in Cisco style:

RP/0/RP0/CPU0:IOSXR#show            <----- so this is top menu as you described, it would give help for completion however entering just this command we don't fall into any menu, it produces error 

% Incomplete command.   <-----
RP/0/RP0/CPU0:IOSXR#show route     <----- if command available we execute and etc 
Thu Jun  8 08:20:22.249 UTC

Codes: C - connected, S - static, R - RIP, B - BGP, (>) - Diversion path
      D - EIGRP, EX - EIGRP external, O - OSPF, IA - OSPF inter area
      N1 - OSPF NSSA external type 1, N2 - OSPF NSSA external type 2
      E1 - OSPF external type 1, E2 - OSPF external type 2, E - EGP
      i - ISIS, L1 - IS-IS level-1, L2 - IS-IS level-2
      ia - IS-IS inter area, su - IS-IS summary null, * - candidate default
      U - per-user static route, o - ODR, L - local, G  - DAGR, l - LISP
      A - access/subscriber, a - Application route
      M - mobile route, r - RPL, t - Traffic Engineering, (!) - FRR Backup path

Gateway of last resort is not set

L    1.1.1.1/32 is directly connected, 00:01:01, Loopback0
C    10.1.0.0/31 is directly connected, 00:00:09, GigabitEthernet0/0/0/0
L    10.1.0.0/32 is directly connected, 00:00:09, GigabitEthernet0/0/0/0
C    10.2.0.0/31 is directly connected, 00:00:09, GigabitEthernet0/0/0/1
L    10.2.0.0/32 is directly connected, 00:00:09, GigabitEthernet0/0/0/1 

People dealing with network devices will intuitively use such a cli  interface

**Reply by daniele77** [reply]
I'd like to understand the value of forbidding access to the menu "show".
I mean: if you only digit "show" at the prompt, you are in the "show" path and then you can choose which item you want to "show" (e.g., "route"). I think it's simply an alternative way with respect to giving the whole command "show route", in my opinion it does not cause any problem.

**Comment by achursin1986** [comment]
That's correct, however my point -  historically this show mode is not available for show commands, it's more for config , when you are at the submode like you type config , then router ospf , you are now at config/ospf , then commands are available for this mode(submenus in terms of the lib).

The value is just mimicking Cisco&Juniper clis, no any functional impact , just user experience option.
